---
title: "Introduction to SurfR"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to SurfR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
## Authors

Aurora Maurizio (maurizio.aurora@hsr.it), Anna Sofia Tascini (tascini.annasofia@unisr.it), Marco Morelli (morelli.marco@hsr.it)



# Introduction

Proteins at the cell surface connect intracellular and extracellular signaling networks and largely determine a cellâ€™s capacity to communicate and interact with its environment. Since transcriptomic differences between healthy-tissue cells and diseased cells might translate into different cell-surface proteins repertoire, the investigation of the cell surfaceome could provide new possibilities for diagnosis, prognosis, treatment development, and therapy response evaluation. Indeed, cell surface proteins i) may act as biomarkers for the early detection of diseased cells in tissues or body fluids and ii) are the most prevalent target of human drugs, as proved by the fact that 66% of approved human drugs listed in the DrugBank database target a cell-surface protein. 

When the goal of a study is to find new biomarkers, the small number of samples and false positive rates are predominant factors limiting the ability to obtain reliable results. However, as costs continue to decrease, it is likely that multiple follow-up experiments will be conducted to re-address some biological questions, suggesting a future need for methods able to jointly analyze data from multiple studies. 

**Surfr**` allows to overcome this issues, generating a list of ranked differentially expressed genes starting from the count matrix of your own RNAseq experiment, or from transcriptomic data automatically retrieved from public databases (e.g. GEO, TCGA).

GEO queries are based on ArchS4 pipeline and data are download in H5 format1. TCGA repository is interrogated through **TCGAbiolinks**. Data retrieved from multiple studies can then be combined through

It also allows to increase the available sample size by integrating related datasets, subsequently increasing the power to detect differential expression. Meta analyisis can therefore be performed through **metaRNASeq** taking into account inter-study variability that may arise due to technical differences among studies (e.g., sample preparation, library protocols, batch effects) as well as additional biological variability. 

The protein classification is based on a recently developed surfaceome predictor, called **SURFY**, based on machine learning3. 

Gene ontology (GO) enrichment analysis, and pathway annotation can also be performed to gain further insights about surface protein candidates of interest.

The package also allows to visualize DEG and enrichment results to help users achieve efficient data interpretation. 

# Installation

SurfR Package can be easily installed from GitHub using devtools:

```
devtools::install_github("addname/SurfR")
```

# Quick Start

The basic idea behind **SurfR** has been to create a complete framework to detect
surface protein coding genes in your data, or public datasets obtained from multiple studies that can be analyzed and compared in a single run, easily revealing functional consensus and differences among distinct conditions.
To begin, let's ask SurfR which to detect surface protein coding genes among a vector of input genes. 
Gene ID can be provided as `gene_name`, `ensembl`, `entrez` or `uniProt_name`.

The protein classification is based on a recently developed surfaceome predictor, called [**SURFY**](http://wlab.ethz.ch/surfaceome/), based on machine learning.


```{r setup}
library(SurfR)
GeneNames = c("CIITA", "EPCAM", "DLK1", "CD24", "CDCP1", "LYVE1", "ABCD1", "VAMP1")
SurfaceProteins_df = Gene2SProtein(GeneNames, input_type = "gene_name")
#The output dataframe contains several information retrieved from Surfy.
colnames(SurfaceProteins_df)
#These are the 5 surface protein coding genes detected by SurfR.
SurfaceProteins_df$UniProt.gene
```



# Tutorial 

## Start from your own data

Although **SurfR** provides many functions to retrieve public data you can always start from your own dataset.

Here we are going to simulate a small bulkRNA dataset with the R package SPsimSeq [REF] 
starting from a subset of Zhang RNA-seq data (Zhang et al. 2015), adding 20% of 
Differentially Expressed genes (`pDE = 0.2`).

You can than decide to stick to it or combine it with other datasets (public or private).

```{r simulate zhang}
library(SPsimSeq)

data("zhang.data.sub")
zhang.counts <- zhang.data.sub$counts
MYCN.status  <- zhang.data.sub$MYCN.status
# Simulation of bulk RNA data with
sim.data.bulk <- SPsimSeq(n.sim = 1, s.data = zhang.counts,
                          group = MYCN.status, n.genes = 1000, batch.config = 1,
                          group.config = c(0.5, 0.5), tot.samples = ncol(zhang.counts),
                          pDE = 0.2, lfc.thrld = 0.5, result.format = "list", return.details = TRUE)

sim.data.bulk1 <- sim.data.bulk$sim.data.list[[1]]

countMatrix = sim.data.bulk1$counts; row.names(countMatrix) <- row.names(zhang.counts)
metadata = sim.data.bulk1$colData; metadata$Group = as.factor(metadata$Group)
```

A fundamental task in the analysis of count data from RNA-seq is the detection 
of differentially expressed genes. For this task we rely on the package DESeq2, 
starting from counts data. The count data are presented as a table which reports, 
for each sample, the number of sequence fragments that have been assigned to each gene.  

We use the built-in SurfR **DGE** function,to perform the differential expression 
analysis of the simulated dataset.

Good Differential Expressed surface protein are the ones which are strongly 
expressed in one condition and almost not expressed in the other. To help detecting those 
candidates, the output data.frame of the DGE function contains information 
on the average expression in the two group (see Mean_CPM_T and Mean_CPM_C columns).

```{r DGE zhang}
library(SurfR)

df <- DGE(expression = countMatrix,
          metadata = metadata,
          Nreplica = 50,
          design = "~Group",
          condition = "Group",
          alpha = 0.05,
          TEST = "1", CTRL =  "0", 
          output_tsv = FALSE)

head(df)
```

Once DEGS have been detected we may want to isolate Surface protein coding genes.

The protein classification is based on a recently developed surfaceome predictor, 
called **SURFY**, based on machine learning. 

We use the built-in SurfR **Gene2SProtein** function to identify
Surface protein coding genes (SP). 

```{r G2P zhang}
# all fdr
fdr_GeneID = df[df$padj < 0.05,"GeneID"]
SP = Gene2SProtein(genes = fdr_GeneID, input_type ="gene_name")

# upregulated fdr
fdrUP_GeneID = df[df$padj < 0.05 & df$log2FoldChange > 0,"GeneID"]
SPup = Gene2SProtein(genes = fdrUP_GeneID, input_type ="gene_name")

# dowregulated fdr
fdrDW_GeneID = df[df$padj < 0.05 & df$log2FoldChange < 0,"GeneID"]
SPdw = Gene2SProtein(genes = fdrDW_GeneID, input_type ="gene_name")
```



## Explore public datasets

## Dowload from GEO (Gene Expression Omnibus)

[GEO](https://www.ncbi.nlm.nih.gov/geo/) is a public functional genomics data repository 
containing high throughput gene expression data and hybridization arrays. 
We provide a handy interface to download experiments and curated gene expression profiles.

Here we are going to reanalyze the [Cloughesy et al](https://doi.org/10.1038/s41591-018-0337-7) 
public dataset of recurrent glioblastoma patients undergoing two different treatments: 
neoadjuvant pembrolizumab or adjuvant pembrolizumab. 

This study is available under the GEO accession series [GSE121810](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE121810).

The metadata is downloaded with SurfR built-in function **GEOmetadata**.
Note that this study has been sequenced with only one sequencing platform.
If this is not the case, you have to download separately all the 
metadata specifying the GPL series numbers, and then merge them.

The counts matrix is downloaded from [ArchS4](https://maayanlab.cloud/archs4/)
with the SurfR built-in function **DownloadArchS4**,
to ensure handling not normalize data.

```{r GEO input}
library(SurfR)
library(stringr)

# Download metadata from GEO
mGSE121810 = GEOmetadata(GSE = "GSE121810")
# create new metadata column in order to remove unwanted special characters
unwanted_character = " "
fx = function(x) {str_split(string = x, pattern = unwanted_character)[[1]][1]}
mGSE121810$condition = sapply(mGSE121810$therapy, fx)
mGSE121810$condition = as.factor(mGSE121810$condition)
# Preview metadata
head(mGSE121810)

# Download count matrix from ArchS4
cGSE121810 = DownloadArchS4(mGSE121810$GSM, species = "human", print_tsv = FALSE, filename = NULL)
# Preview count matrix
head(cGSE121810[,1:5])
```

A fundamental task in the analysis of count data from RNA-seq is the detection 
of differentially expressed genes. For this task we rely on the package DESeq2, 
starting from counts data. The count data are presented as a table which reports, 
for each sample, the number of sequence fragments that have been assigned to each gene.  

We use the built-in SurfR **DGE** function,to perform the differential expression 
analysis of the GSE121810 dataset.

Good Differential Expressed surface protein are the ones which are strongly 
expressed in one condition and almost not expressed in the other. To help detecting those 
candidates, the output data.frame of the DGE function contains information 
on the average expression in the two group (see Mean_CPM_T and Mean_CPM_C columns).

```{r GEO dge}
# Perform DGE
df <- DGE(expression = cGSE121810,
          metadata = mGSE121810,
          Nreplica = 14,
          design = "~condition",
          condition = "condition",
          alpha = 0.05,
          TEST = "neoadjuvant", CTRL =  "adjuvant",
          output_tsv = F)

head(df)
```

Once DEGS have been detected we may want to isolate Surface protein coding genes.

The protein classification is based on a recently developed surfaceome predictor, 
called **SURFY**, based on machine learning. 

We use the built-in SurfR **Gene2SProtein** function to identify
Surface protein coding genes (SP). 

```{r GEO SP}
# Detect SP amoung differentially expressed genes
fdr_GeneID = df[df$padj < 0.05,"GeneID"]
SP = Gene2SProtein(genes = fdr_GeneID, input_type ="gene_name")

fdrUP_GeneID = df[df$padj < 0.05 & df$log2FoldChange > 0,"GeneID"]
SPup = Gene2SProtein(genes = fdrUP_GeneID, input_type ="gene_name")

fdrDW_GeneID = df[df$padj < 0.05 & df$log2FoldChange < 0,"GeneID"]
SPdw = Gene2SProtein(genes = fdrDW_GeneID, input_type ="gene_name")

```


## Dowload from TCGA

The Cancer Genome Atlas (TCGA, https://tcga-data.nci.nih.gov/tcga/) contains data 
for thousands of tumor samples across more than 20 types of cancer.
Navigating through all of the files manually is impossible. 
Therefore we provide a function based on TCGAbiolinks that automates and 
streamlines the retrieval of public TCGA transcriptomics data.
Note that to use this function you need to install the developmental version of
[TCGAbiolink](https://github.com/BioinformaticsFMRP/TCGAbiolinks).

```
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("BioinformaticsFMRP/TCGAbiolinksGUI.data")
BiocManager::install("BioinformaticsFMRP/TCGAbiolinks")
```

Here we are going to reanalyze the *TCGA-THYM* dataset, since is one of the smallest 
TCGA datasets which also includes two normal samples. 

The TCGA count matrix and the metadata are downloaded with SurfR built-in 
function **TCGA_download**. The shortLetterCode column of the metadata allow us 
to distinguish Primary Solid Tumor (TP) and normal (NT) samples.


```{r TCGA input}
library(SurfR)
# Download TCGA data
TCGA.THYM = TCGA_download(project = "TCGA-THYM")
cTCGA.THYM <- TCGA.THYM[[1]]
mTCGA.THYM <- TCGA.THYM[[2]]
table(mTCGA.THYM$shortLetterCode)

mTCGA.THYM$shortLetterCode = as.factor(mTCGA.THYM$shortLetterCode)
```


A fundamental task in the analysis of count data from RNA-seq is the detection 
of differentially expressed genes. For this task we rely on the package DESeq2, 
starting from counts data. The count data are presented as a table which reports, 
for each sample, the number of sequence fragments that have been assigned to each gene.  

We use the built-in SurfR **DGE** function,to perform the differential expression 
analysis of the TCGA-THYM dataset.

Good Differential Expressed surface protein are the ones which are strongly 
expressed in one condition and almost not expressed in the other. To help detecting those 
candidates, the output data.frame of the DGE function contains information 
on the average expression in the two group (see Mean_CPM_T and Mean_CPM_C columns).

```{r TCGA DGE}

df <- DGE(expression = cTCGA.THYM,
          metadata = mTCGA.THYM,
          Nreplica = 2,
          design = "~shortLetterCode",
          condition = "shortLetterCode",
          alpha = 0.05,
          TEST = "TP", CTRL =  "NT",
          output_tsv = F)

head(df)
```


Once DEGS have been detected we may want to isolate Surface protein coding genes.

The protein classification is based on a recently developed surfaceome predictor, 
called **SURFY**, based on machine learning. 

We use the built-in SurfR **Gene2SProtein** function to identify
Surface protein coding genes (SP).

```{r TCGA SP}
fdr_GeneID = df[df$padj < 0.05,"GeneID"]
SP = Gene2SProtein(genes = fdr_GeneID, input_type ="gene_name")

fdrUP_GeneID = df[df$padj < 0.05 & df$log2FoldChange > 0,"GeneID"]
SPup = Gene2SProtein(genes = fdrUP_GeneID, input_type ="gene_name")

fdrDW_GeneID = df[df$padj < 0.05 & df$log2FoldChange < 0,"GeneID"]
SPdw = Gene2SProtein(genes = fdrDW_GeneID, input_type ="gene_name")
```



# Perform Meta-analisys

Analyzing data arising from several experiments studying the same question is a way to obtain more robust results, increasing the detection power of differentially expressed genes. 

## Venn diagram

A DEGs analysis can result in many upregulated genes and it may be difficult to choose the best surface marker candidatates. If you analyzed several datasets we always suggest to start taking a look at the DEGs in common beetween different sources with a simple, yet effective representation strategy such as the venn diagram.

```
library(SurfR)
```

## pvalue combination

Here we provide a function based on **metaRNASeq** bioconductor package to implement two p-value combination techniques (inverse normal and Fisher methods).


## inverse normal

```
library(SurfR)
```

## Fisher


```
library(SurfR)
```


# Functional Enrichment
Once we determined the subset of genes enriched in our condition of interest, we have multiple analysis to perform to go beyond a simple list of genes. 

We can perform a general enrichment analysis to gain further insights about upregulated or downregulated DEGs.

```
library(SurfR)
```
We can annotate our list of genes with cross-databases identifiers and descriptions (Entrezid, Uniprot, KEGG, etc.) taking advantage of one the 35 gene-set libraries present in the Enrichr database.

```
library(SurfR)
```


# Visualization of functional enrichment result
**SurfR** implements several visualization methods to help interpreting enrichment results obtained through EnrichR (Kuleshov et al. 2013) using ggplot2.


```
library(SurfR)
```


### Bar plot 

It depicts and gene count ratio and enrichment scores (- Log10 adjusted p values) as bar height and color. Users can specify the number of terms (most significant) to display.


```
library(SurfR)
```

### Dot plot 

Dot plot is similar to bar plot with the capability to encode another score as dot size.


```
library(SurfR)
```



