---
title: "Introduction to SurfR"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to SurfR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
## Authors

Aurora Maurizio (maurizio.aurora@hsr.it), Anna Sofia Tascini (tascini.annasofia@unisr.it), Marco Morelli (morelli.marco@hsr.it)



# Introduction

Proteins at the cell surface connect intracellular and extracellular signaling networks and largely determine a cellâ€™s capacity to communicate and interact with its environment. Since transcriptomic differences between healthy-tissue cells and diseased cells might translate into different cell-surface proteins repertoire, the investigation of the cell surfaceome could provide new possibilities for diagnosis, prognosis, treatment development, and therapy response evaluation. Indeed, cell surface proteins i) may act as biomarkers for the early detection of diseased cells in tissues or body fluids and ii) are the most prevalent target of human drugs, as proved by the fact that 66% of approved human drugs listed in the DrugBank database target a cell-surface protein. 

When the goal of a study is to find new biomarkers, the small number of samples and false positive rates are predominant factors limiting the ability to obtain reliable results. However, as costs continue to decrease, it is likely that multiple follow-up experiments will be conducted to re-address some biological questions, suggesting a future need for methods able to jointly analyze data from multiple studies. 

**Surfr**` allows to overcome this issues, generating a list of ranked differentially expressed genes starting from the count matrix of your own RNAseq experiment, or from transcriptomic data automatically retrieved from public databases (e.g. GEO, TCGA).

GEO queries are based on ArchS4 pipeline and data are download in H5 format1. TCGA repository is interrogated through **TCGAbiolinks**. Data retrieved from multiple studies can then be combined through

It also allows to increase the available sample size by integrating related datasets, subsequently increasing the power to detect differential expression. Meta analyisis can therefore be performed through **metaRNASeq** taking into account inter-study variability that may arise due to technical differences among studies (e.g., sample preparation, library protocols, batch effects) as well as additional biological variability. 

The protein classification is based on a recently developed surfaceome predictor, called **SURFY**, based on machine learning3. 

Gene ontology (GO) enrichment analysis, and pathway annotation can also be performed to gain further insights about surface protein candidates of interest.

The package also allows to visualize DEG and enrichment results to help users achieve efficient data interpretation. 

# Quick Start

The basic idea behind **SurfR** has been to create a complete framework to detect
surface protein coding genes in your data, or public datasets obtained from multiple studies that can be analyzed and compared in a single run, easily revealing functional consensus and differences among distinct conditions.
To begin, we load the necessary packages and simulation parameters:

```{r setup}
library(SurfR)
```

# Quick Start

The basic idea behind **SurfR** has been to create a complete framework to detect
surface protein coding genes in your data, or public datasets obtained from multiple studies that can be analyzed and compared in a single run, easily revealing functional consensus and differences among distinct conditions.
To begin, let's ask SurfR which to detect surface protein coding genes among a vector of input genes. Gene names can be provided as \code{gene_name}, \code{ensembl}, \code{entrez} or \code{uniProt_name}.
The protein classification is based on a recently developed surfaceome predictor, called [**SURFY**](http://wlab.ethz.ch/surfaceome/), based on machine learning .


```{r setup}
library(SurfR)
GeneNames = c("CIITA", "EPCAM", "DLK1", "CD24", "CDCP1", "LYVE1", "ABCD1", "VAMP1")
SurfaceProteins_df = Gene2SProtein(GeneNames, input_type = "gene_name")
#The output dataframe contains several information retrieved from Surfy.
colnames(SurfaceProteins_df)
#These are the 5 surface protein coding genes detected by SurfR.
SurfaceProteins_df$UniProt.gene
```



# Tutorial 

## Start from your own data

Although **SurfR** provides many functions to retrieve public data you can always start from your own dataset.
Here we are going to make an example with the Differentiation of Human Embryonic Stem Cells under Hypoxia gene
expression dataset by Prado-Lopez et al. (2010). 
For simplicity we are going to use only a subset of it focusing on 2 sample processing time points (0 and 15) and 
and only one level of oxygen available in the conditioned medium (21).
You can than decide to stick to it or combine it with other datasets (public or private).

```
library(SurfR)

BiocManager::install("tweeDEseqCountData")
library(tweeDEseqCountData)
data(pickrell1)
data(montgomery)
montgomery.eset$
devtools::install_github("zhangyuqing/sva-devel")

library(sva)
#count_matrix <- matrix(rnbinom(400, size=10, prob=0.1), nrow=50, ncol=8)
#batch <- c(rep(1, 4), rep(2, 4))
#adjusted <- ComBat_seq(count_matrix, batch=batch, group=NULL)


head(count_matrix)

head(pickrell1.eset$featureData$fvarMetadata)
Counts <- exprs(pickrell1.eset)
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
  

BiocManager::install("stemHypoxia")
library(stemHypoxia)
data(stemHypoxia)
## Exploring the expression matrix M
dim(M) #40736 features x 28 sample + ID + Symbol
head(M) #Just to see some intensity values
colnames(M)
 head(design)
SurfaceProteins_df = Gene2SProtein(M$GeneName, input_type = "gene_name")

# ------- input data -------------
# remove duplicated entry from GeneName column
M <- M[!duplicated(M$GeneName), ]

# filter for chosen timepoint and oxygen
designf = design[design$time %in% c(0.0, 15.0) & design$oxygen == 21,]
designf$time <- as.factor(designf$time)
Mf = M[, designf$samplename]
row.names(Mf) <- M$GeneName

row.names(designf) = designf$samplename

# ------- Differential Gene Expression -------------

EG1_df = DGE(Mf, designf, design = "~time", TEST = "0.0", CTRL = "15.0", Nreplica = 2, alpha = 0.05, output_tsv = F)

```
## Explore public datasets



## Dowload from GEO (Gene Expression Omnibus)

GEO is a public functional genomics data repository containg high throughput gene expression data and hybridization arrays. We provide a handy interface to download experiments and curated gene expression profiles.


```
library(SurfR)
```



## Dowload from TCGA
The Cancer Genome Atlas (TCGA, https://tcga-data.nci.nih.gov/tcga/) contains data for thousands of tumor samples across more than 20 types of cancer.
Navigating through all of the files manually is impossible. Therefore we provide a function based on TCGAbiolinks that automates and streamlines the retrieva of public TCGA transcriptomicsdata.



```
library(SurfR)
```


## Start from your own data

Although **SurfR** provides many functions to retrieve public data you can always start from your own dataset.
You can than decide to stick to it or combine it with other datasets (public or private).



```
library(SurfR)
```


## DEG

An fundamental task in the analysis of count data from RNA-seq is the detection of differentially expressed genes. For this task we rely on the package DESeq2, starting from counts data.
The count data are presented as a table which reports, for each sample, the number of sequence fragments that have been assigned to each gene.  


### Example with counts data retrieved from GEO

```
library(SurfR)
```

### Example with counts data retrieved from TCGA
```
library(SurfR)
```

## Identify surface proteins

Once DEGS have been detected we may want to isolate Surface protein coding genes.
The protein classification is based on a recently developed surfaceome predictor, called **SURFY**, based on machine learning. 

```
library(SurfR)
```


## Perform Meta-analisys

Analyzing data arising from several experiments studying the same question is a way to obtain more robust results, increasing the detection power of differentially expressed genes. 

### Venn diagram

A DEGs analysis can result in many upregulated genes and it may be difficult to choose the best surface marker candidatates. If you analyzed several datasets we always suggest to start taking a look at the DEGs in common beetween different sources with a simple, yet effective representation strategy such as the venn diagram.

```
library(SurfR)
```

### pvalue combination

Here we provide a function based on **metaRNASeq** bioconductor package to implement two p-value combination techniques (inverse normal and Fisher methods).


### inverse normal

```
library(SurfR)
```

### Fisher


```
library(SurfR)
```


## Functional Enrichment
Once we determined the subset of genes enriched in our condition of interest, we have multiple analysis to perform to go beyond a simple list of genes. 

We can perform a general enrichment analysis to gain further insights about upregulated or downregulated DEGs.

```
library(SurfR)
```
We can annotate our list of genes with cross-databases identifiers and descriptions (Entrezid, Uniprot, KEGG, etc.) taking advantage of one the 35 gene-set libraries present in the Enrichr database.

```
library(SurfR)
```


## Visualization of functional enrichment result
**SurfR** implements several visualization methods to help interpreting enrichment results obtained through EnrichR (Kuleshov et al. 2013) using ggplot2.


```
library(SurfR)
```


### Bar plot 

It depicts and gene count ratio and enrichment scores (- Log10 adjusted p values) as bar height and color. Users can specify the number of terms (most significant) to display.


```
library(SurfR)
```

### Dot plot 

Dot plot is similar to bar plot with the capability to encode another score as dot size.


```
library(SurfR)
```



